addon/pods/components/rui-validatable-input/component.js

```diff
@@ -8,35 +8,103 @@ const {
 } = Ember
 
 export default Component.extend({
-  classNameBindings: ['showErrorClass:has-error', 'isValid:has-success'],
+  classNameBindings: ['isInvalid:has-error', 'isValid:has-success'],
   classNames: ['rui-validatable-input'],
   layout,
-  model: null,
-  placeholder: '',
+
+  // Default Attrs
+
   type: 'text',
-  validation: null,
-  value: null,
-  valuePath: '',
+  async: false,
+
+  // Passed Attrs
+
+  // valuePath: string
+  // model: object
+
+  // Init Attrs
+
+  // value: @alias model.valuePath
+  // validation: @oneway model.validations.attrs.valuePath
 
   init() {
     this._super(...arguments)
     const valuePath = this.get('valuePath')
-    defineProperty(this, 'validation', computed.oneWay(`model.validations.attrs.${valuePath}`))
+    defineProperty(this,
+      'validation',
+      computed.oneWay(`model.validations.attrs.${valuePath}`))
     defineProperty(this, 'value', computed.alias(`model.${valuePath}`))
   },
 
-  didChange: computed('value', 'model.hasDirtyAttributes', function() {
-    let attrsChanged = this.get('model') ? this.get('model').changedAttributes() : []
-    return this.get('valuePath') in attrsChanged
+  // Validation binding attributes with check for async settings
+  // if async is true, don't display either until `_didValidate`
+
+  isInvalid: computed('_isInvalid', '_didValidate', function() {
+    if (this.get('async')) {
+      return this.get('_isInvalid') && this.get('_didValidate')
+    }
+    return this.get('_isInvalid')
+  }),
+
+  isValid: computed('_isValid', '_didValidate', function() {
+    if (this.get('async')) {
+      return this.get('_isValid') && this.get('_didValidate')
+    }
+    return this.get('_isValid')
   }),
-  didValidate: computed.oneWay('targetObject.didValidate'),
-  hasContent: computed.notEmpty('value'),
-  isInvalid: computed.oneWay('validation.isInvalid'),
-  isValid: computed.and('didChange', 'hasContent', 'validation.isValid', 'notValidating'),
-  notValidating: computed.not('validation.isValidating'),
-  showErrorClass: computed.and('notValidating', 'showMessage', 'validation'),
-  showMessage: computed('validation.isDirty', 'isInvalid', 'didValidate', function() {
-    return (this.get('validation.isDirty') || this.get('didValidate')) && this.get('isInvalid')
-  })
 
+  // This method ensures that if presence is required
+  // But the initial value is undefined or an empty string
+  // (the case with new records) we can still validate even though
+  // the `_didChange` attr would report it had not
+
+  _validatePresenceWithEmptyDefault: computed('value', function() {
```

Comment:
the naming of this confuses me a lot...I kinda know and am still struggling to get my head around it.